src/VisualStudio/Core/Def/Implementation/Snippets/SnippetFunctions/AbstractSnippetFunctionArgumentValue.cs

```diff
@@ -0,0 +1,101 @@
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using VsTextSpan = Microsoft.VisualStudio.TextManager.Interop.TextSpan;
+using static Roslyn.Utilities.TaskExtensions;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Snippets
+{
+    internal abstract class AbstractSnippetFunctionArgumentValue : AbstractSnippetFunction
+    {
+        /// <summary>
+        /// The name of the parameter for which an argument value is requested.
+        /// </summary>
+        private readonly string _parameterName;
+
+        /// <summary>
+        /// A <see cref="SymbolKey"/> allowing the <see cref="IParameterSymbol"/> to be resolved.
+        /// </summary>
+        private readonly SymbolKey _parameterKey;
+
+        protected AbstractSnippetFunctionArgumentValue(AbstractSnippetExpansionClient snippetExpansionClient, ITextBuffer subjectBuffer, string parameterName, SymbolKey parameterKey)
+            : base(snippetExpansionClient, subjectBuffer)
+        {
+            _parameterName = parameterName;
+            _parameterKey = parameterKey;
+        }
+
+        protected abstract string? FallbackDefaultLiteral { get; }
+
+        protected override int GetDefaultValue(CancellationToken cancellationToken, out string? value, out int hasCurrentValue)
+        {
+            value = snippetExpansionClient.ThreadingContext.JoinableTaskFactory.Run(() => TryGetDefaultValueAsync(cancellationToken));
+            if (value is not null)
+            {
+                hasCurrentValue = 1;
+                return VSConstants.S_OK;
+            }
+            else
+            {
+                hasCurrentValue = 0;
+                return VSConstants.E_FAIL;
+            }
+        }
+
+        [SuppressMessage("Reliability", "CA2007:Consider calling ConfigureAwait on the awaited task", Justification = "Containing method uses JTF https://github.com/dotnet/roslyn-analyzers/issues/4283")]
+        private async Task<string?> TryGetDefaultValueAsync(CancellationToken cancellationToken)
+        {
+            await snippetExpansionClient.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            if (!TryGetDocument(out var document))
+            {
+                // Unable to suggest arguments if a document is not available
+                return null;
+            }
+
+            var textSpan = new VsTextSpan[1];
+            if (snippetExpansionClient.ExpansionSession.GetFieldSpan(_parameterName, textSpan) != VSConstants.S_OK)
+            {
+                // Failed to obtain the snippet placeholder for the named parameter.
+                // TODO: Telemetry https://github.com/dotnet/roslyn/issues/50033
+                return null;
+            }
+
+            var compilation = await document.Project.GetRequiredCompilationAsync(cancellationToken);
```

Comment:
very very nervous about the implications here.  specifically, the UI affinitized expansion system, but all this potentially expensive semantic work.

That said, this is an experiemental feature, so this is non blocking.  just something for us to be cautious about.

Reply:
➡️ Now has a TODO and https://github.com/dotnet/roslyn/issues/50634